src/SqlSerializer.cpp

```diff
@@ -499,66 +610,134 @@ sqlite3_stmt *SqlSerializer::populate_promise_association_statement(
+}
+
+void SqlSerializer::clear_verbose_information() { verbose_information.clear(); }
+
+size_t SqlSerializer::verbose_information_count() {
+    return verbose_information.size();
+}
+
 sqlite3_stmt *
 SqlSerializer::populate_insert_argument_statement(const closure_info_t &info,
                                                   int index) {
     const arg_t &argument = info.arguments.all()[index].get();
-    sqlite3_bind_int(insert_argument_statement, 1, get<1>(argument));
-    sqlite3_bind_text(insert_argument_statement, 2, get<0>(argument).c_str(),
-                      -1, SQLITE_TRANSIENT);
-    sqlite3_bind_int(insert_argument_statement, 3,
-                     index); // FIXME broken or unnecessary (pick one)
-    sqlite3_bind_int(insert_argument_statement, 4, info.call_id);
+    bind_int(insert_argument_statement, 1, get<1>(argument));
+    bind_text(insert_argument_statement, 2, get<0>(argument).c_str(), -1,
+              SQLITE_TRANSIENT);
+    bind_int(insert_argument_statement, 3,
+             index); // FIXME broken or unnecessary (pick one)
+    bind_int(insert_argument_statement, 4, info.call_id);
     return insert_argument_statement;
 }
+
+int SqlSerializer::bind_int(sqlite3_stmt *statement, int parameter_index,
+                            int value, std::string annotation) {
+    if (verbose) {
+        add_verbose_information(
+            sqlite3_bind_parameter_name(statement, parameter_index) + 1,
+            std::to_string(value).length(), annotation);
+    }
+    return sqlite3_bind_int(statement, parameter_index, value);
+}
+
+int SqlSerializer::bind_int64(sqlite3_stmt *statement, int parameter_index,
+                              sqlite3_int64 value, std::string annotation) {
+    if (verbose) {
+        add_verbose_information(
+            sqlite3_bind_parameter_name(statement, parameter_index) + 1,
+            /* TODO - check if to_string will behave
+               correctly on sqlite3_int64 */
+            std::to_string(value).length(), annotation);
+    }
+    return sqlite3_bind_int64(statement, parameter_index, value);
+}
+
+int SqlSerializer::bind_double(sqlite3_stmt *statement, int parameter_index,
+                               double value, std::string annotation) {
+    if (verbose) {
+        add_verbose_information(
+            sqlite3_bind_parameter_name(statement, parameter_index) + 1,
+            std::to_string(value).length(), annotation);
+    }
+    return sqlite3_bind_double(statement, parameter_index, value);
+}
+
+int SqlSerializer::bind_null(sqlite3_stmt *statement, int parameter_index,
+                             std::string annotation) {
+    if (verbose) {
+        add_verbose_information(
+            sqlite3_bind_parameter_name(statement, parameter_index) + 1, 4,
+            annotation);
+    }
+    return sqlite3_bind_null(statement, parameter_index);
+}
+
+int SqlSerializer::bind_text(sqlite3_stmt *statement, int parameter_index,
```

Comment:
why not this? so much safer (memory managed and type checked, i think)
```
int bind_text(sqlite3_stmt *, int, string const & str) {
... str.c_str() ...
}
```

Reply:
I want to mirror the exact interface of `sqlite3_bind_*` functions. The only change is an extra annotation parameter which defaults to empty string. What you are suggesting can be done, but then we will also have to update all arguments passed to this parameter in the `SqlSerializer` class. Its a small change, but it will make this function slightly different from others which exactly mimic their `sqlite3_bind_*` counterparts.

Reply:
If this is only about verisimilitude, I personally think we should stick to C++ idioms, since they provide fewer chances to screw up pointers.